File: go/sketchkit/internal/hashfamily/hashfamily.go
// Code authors: Vijay Erramilli and Codex
// Package hashfamily implements the deterministic sketch-local hash family
// defined in spec/hash.md. It is not a keyed hash and must only be applied to
// already-keyed uint64 sketch inputs.
package hashfamily

const (
	BloomSeed   uint64 = 0x83984a98fd448a39
	MinHashSeed uint64 = 0xea59f2718f8069a6
)

// Mix64 is the spec-defined 64-bit SplitMix finalizer.
func Mix64(x uint64) uint64 {
	x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9
	x = (x ^ (x >> 27)) * 0x94d049bb133111eb
	return x ^ (x >> 31)
}

// BloomHash returns the i-th Bloom hash-family value for hash.
func BloomHash(hash uint64, index uint32) uint64 {
	return Mix64(hash ^ Mix64(BloomSeed^uint64(index)))
}

// MinHash returns the i-th MinHash hash-family value for hash.
func MinHash(hash uint64, index uint32) uint64 {
	return Mix64(hash ^ Mix64(MinHashSeed^uint64(index)))
}

File: python/llm_sketchkit/hashfamily.py
# Code authors: Vijay Erramilli and Codex
"""Deterministic sketch-local hash families from `spec/hash.md`."""

from __future__ import annotations

import hashlib

MASK64 = (1 << 64) - 1
BLOOM_TAG = b"llm-sketchkit:bloom:v1"
MINHASH_TAG = b"llm-sketchkit:minhash:v1"

BLOOM_SEED = int.from_bytes(hashlib.sha256(BLOOM_TAG).digest()[:8], "big")
MINHASH_SEED = int.from_bytes(hashlib.sha256(MINHASH_TAG).digest()[:8], "big")


def mix64(value: int) -> int:
    """Return the SplitMix finalizer with unsigned 64-bit overflow."""

    x = value & MASK64
    x = ((x ^ (x >> 30)) * 0xBF58476D1CE4E5B9) & MASK64
    x = ((x ^ (x >> 27)) * 0x94D049BB133111EB) & MASK64
    return (x ^ (x >> 31)) & MASK64


def bloom_hash(value: int, index: int) -> int:
    """Return the Bloom hash-family value for a keyed sketch hash."""

    seed = mix64(BLOOM_SEED ^ (index & MASK64))
    return mix64(value ^ seed)


def minhash(value: int, index: int) -> int:
    """Return the MinHash hash-family value for a keyed sketch hash."""

    seed = mix64(MINHASH_SEED ^ (index & MASK64))
    return mix64(value ^ seed)
